### DISCUSSION TITLE
Adding a Webhook that works like Freeform's POST Forwarding feature?

### DISCUSSION THREAD

**Discussion by chrismlusk** [verbb/formie#1361] [Q&A] [answered]
I'm trying to add a Webhook integration to forward form submissions to Oracle Eloqua. I've created the webhook, set the URL, and added it to my test form. But, when I submit the form, I get this error:

```
API error: “Syntax error” /var/www/html/app/vendor/yiisoft/yii2/helpers/BaseJson.php:147
```

I assume the error is because the response from Eloqua is not JSON. This is the response I see in my local API client:

```html
<script type='text/javascript'>
	document.location.href = 'https://s61956.t.eloqua.com/e/DefaultFormSubmitConfirmation.aspx';
</script>
```

I am attempting to migrate a project to Formie from Freeform. The lead form in this project currently uses Freeform's [POST forwarding](https://docs.solspace.com/craft/freeform/v4/integrations/post-forwarding/) feature to send data to the same Eloqua form. I was hoping I could use Formie webhooks to accomplish the same thing, but is that not the case?

**Comment by engram-design** [comment, marked as answer]
That's correct the Webhook integration currently expects a JSON response. Although the Webhook integration essentially otherwise acts like POST forwarding in Freeform.

In addition, something like this won't typically be possible with Formie, because by default integrations are run via the queue. If you're wanting to redirect the user to another page, that won't really work in the context of a queue job, which might run after the actual form submission.

There's no way to handle that apart from setting the redirect URL on the form, or implementing some module code to set the redirect URL (if this needs to be dynamic, and in context of the form, etc).

**Reply by chrismlusk** [reply]
In this particular case, I don't think I'm worried about the queue because I do want Formie to handle everything as it is designed to (submission, redirect, notification, etc.) — I just want to _also_ forward the submitted data to this third-party service. The forwarding doesn't even need to be immediate, so maybe a custom integration would work here?

I'll need to get more familiar and learn a bit before setting off on that path, but do you think that would be a solution for me to continue the switch to Formie?

**Reply by engram-design** [reply]
No problem, I just saw your response back using `document.location.href` which won't work for Formie - just in case you had something setup where the response of the Webhook changed the redirect behaviour in Freeform.

I've actually just changed this behaviour for you with the Webhook integration. We now no longer enforce a JSON response, just a HTTP 20x response. To get this early, run `composer require verbb/formie:"dev-craft-4 as 2.0.24"`.

**Reply by chrismlusk** [reply]
That's helpful. Thank you! 

I think I still might have to explore a custom integration to get the payload to match what the service is expecting, because I'm now seeing the following:

```html
Success!
<table align="center">
<tr>
<td style="FONT-SIZE: 10pt; WIDTH: 400px; PADDING-TOP: 10px; FONT-FAMILY: Arial, Helevtica" align="left">
<p>
A problem has occurred with this form. The system administrator has been notified and the problem will be addressed immediately.
</p>
<p>Thank you for your patience.</p>
</td>
</tr>
</table>
```

This Oracle service is frustrating because there's pretty much nothing I can do to configure anything on that end, AFAIK, but I think I can do what I need on the Craft side with Formie. We'll see.

**Reply by engram-design** [reply]
Ah gotcha, yeah if you're trying to send it to an endpoint or service you don't have control over, you'll likely want a custom integration to tailor that. I did assume you had control over the endpoint. It probably differs because we send the payload as JSON, where Freeform (and your endpoint) might expect a raw POST request payload.

If it helps at all, there's a [user guide](https://verbb.io/craft-plugins/formie/user-guides/building-a-webhook-integration-from-scratch) on creating a custom integration.

**Reply by chrismlusk** [reply]
Yep, creating a custom Miscellaneous integration that sends a raw POST request did the trick. And thanks to the plugin's thoughtfully designed API, doing this was so incredibly simple. Cheers for building such a delightful product!
